### ISSUE TITLE
Login-module breaks 404-Error-page

### ISSUE THREAD

**Issue by asaage** [contao/contao#999]
**Affected version(s)**
4.8.5 and probably below

**Description**
Loginmodule on Error404 page renders ugly **Oops! An Error Occurred** symfony-twig-template
https://community.contao.org/de/showthread.php?76129-404-Seiten
That happens with misspelled page-aliases or url-suffixes but funnily enough not wit wrong auto_item parameters. Took me ages to figure that out as neither logfiles, the console route-match command nor the debug-mode was very helpfull.

**How to reproduce**  
Put a Login-module on the 404-Error-page (layout or content - doesn't matter) and then abuse your address-bar. (i reproduced it in the online-demo)

**Comment by fritzmg** [comment]
I can confirm the problem.

**Comment by aschempp** [comment]
any idea what the actual error message is (try running in debug mode)?

**Comment by asaage** [comment]
Symfony\Component\Routing\Exception\ResourceNotFoundException

None of the routers in the chain matched this request GET /foo.html

**Comment by aschempp** [comment]
That's the message in the login module "frame" ??

**Comment by fritzmg** [comment]
![capture](https://user-images.githubusercontent.com/4970961/69228183-67e67800-0b83-11ea-8b56-f64166044076.png)

```
Symfony\Component\HttpKernel\Exception\NotFoundHttpException:
No route found for "GET /foo"

  at vendor/symfony/http-kernel/EventListener/RouterListener.php:138
  at Symfony\Component\HttpKernel\EventListener\RouterListener->onKernelRequest(object(RequestEvent), 'kernel.request', object(EventDispatcher))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:298)
  at Symfony\Component\EventDispatcher\EventDispatcher::Symfony\Component\EventDispatcher\{closure}(object(RequestEvent), 'kernel.request', object(EventDispatcher))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:260)
  at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure)), 'kernel.request', object(RequestEvent))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:235)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners(array(object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure)), 'kernel.request', object(RequestEvent))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:73)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch(object(RequestEvent), 'kernel.request')
     (vendor/symfony/http-kernel/HttpKernel.php:127)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:198)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/index.php:31)
  at require('...\\c48dev\\web\\index.php')
     (web/app.php:4)
```

```
Symfony\Component\Routing\Exception\ResourceNotFoundException:
None of the routers in the chain matched this request
GET /foo HTTP/1.1
Accept:                    text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Encoding:           gzip, deflate
Accept-Language:           en-GB,en;q=0.8,de-AT;q=0.5,de;q=0.3
Connection:                keep-alive
Content-Length:            
Content-Type:              
Cookie:                    PHPSESSID=fnm9cjbk5on3hqpno3cfoktlnm; csrf_contao_csrf_token=du3McI2BY2D_Nw-oF05MFH_mNWN4DvqT3G8baEeb0dA
Host:                      c48.local
Upgrade-Insecure-Requests: 1
User-Agent:                Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:70.0) Gecko/20100101 Firefox/70.0
X-Php-Ob-Level:            1
Cookie: PHPSESSID=fnm9cjbk5on3hqpno3cfoktlnm; csrf_contao_csrf_token=du3McI2BY2D_Nw-oF05MFH_mNWN4DvqT3G8baEeb0dA



  at vendor/symfony-cmf/routing/src/ChainRouter.php:210
  at Symfony\Cmf\Component\Routing\ChainRouter->doMatch('/foo', object(Request))
     (vendor/symfony-cmf/routing/src/ChainRouter.php:158)
  at Symfony\Cmf\Component\Routing\ChainRouter->matchRequest(object(Request))
     (vendor/symfony/http-kernel/EventListener/RouterListener.php:114)
  at Symfony\Component\HttpKernel\EventListener\RouterListener->onKernelRequest(object(RequestEvent), 'kernel.request', object(EventDispatcher))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:298)
  at Symfony\Component\EventDispatcher\EventDispatcher::Symfony\Component\EventDispatcher\{closure}(object(RequestEvent), 'kernel.request', object(EventDispatcher))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:260)
  at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure)), 'kernel.request', object(RequestEvent))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:235)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners(array(object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure), object(Closure)), 'kernel.request', object(RequestEvent))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:73)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch(object(RequestEvent), 'kernel.request')
     (vendor/symfony/http-kernel/HttpKernel.php:127)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:68)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:198)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/index.php:31)
  at require('...\\c48dev\\web\\index.php')
     (web/app.php:4)
```

**Comment by asaage** [comment]
What do you mean with "frame"?
![image](https://user-images.githubusercontent.com/4138336/69228410-cf042c80-0b83-11ea-9150-70ad14c95eb3.png)

**Comment by leofeyer** [comment]
As I have commented in https://github.com/contao/contao/issues/967#issuecomment-555956912, I do not consider it a good practice to have a login module on an error site.

**Comment by fritzmg** [comment]
@leofeyer that would mean you would always need to use a custom page layout for the 404, 403 and 401 pages, if you have a login module in your page layout? I do not think this should be necessary.

Also I think having a login module on the 401 page is definitely something that should be supported.

**Comment by asaage** [comment]
I partially agree on "good practice" especially if the login has no JumpTo.
What's the quickest was to get the page-type in an arbitary template?
I would add a condition (dont realy insist on that module being present).
Additional layouts are always a hassle for me. 🤔

**Comment by richardhj** [comment]
The error page gets rendered by the PrettyErrorScreenListener.
When placing a login form on the 404 page, another exception gets thrown: [the AuthenticationCredentialsNotFoundException](https://github.com/symfony/symfony/blob/fa358e61e735b335ae0a9815dd9955a019a3d01f/src/Symfony/Component/Security/Core/Authorization/AuthorizationChecker.php#L49) because no token is in the token storage.

I don't know if we can load the token storage correctly.
Otherwise we would need to catch that exception in the login module and return an empty respone.

**Comment by aschempp** [comment]
Can someone test if this still happens in the latest master branch?

**Comment by richardhj** [comment]
Yep, issue persists.
It is caused by https://github.com/contao/contao/blob/275dd6a65744c09547a7983688b9b269c11f4b21/core-bundle/src/Resources/contao/modules/ModuleLogin.php#L102 which will throw the AuthenticationCredentialsNotFoundException due to the missing token.

**Comment by aschempp** [comment]
I was very confused by the fact that the `AuthorizationChecker` throws an exception without a token. Then I realized that we normally *always* have a token, at least an *Anonymous* one if a user is not logged in.

Unfortunately, the router has higher listener priority than the security firewall, and an exception throws in the router will bypass the firewall listener, so no anonymous token is set. The best fix I can think of is to make sure there is an anonymous token on the 404 error page. However, this means a user is **always** not logged in on the 404 page. I think we discussed that before, the only alternative would be to (again) have a catch-all route that triggers the 404 instead of letting the router generate a 404 exception.
